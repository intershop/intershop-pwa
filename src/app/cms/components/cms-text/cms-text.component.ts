import { ChangeDetectionStrategy, Component, Input } from '@angular/core';

import { ContentPageletView } from 'ish-core/models/content-view/content-views';
import { SfeMetadataWrapper } from '../../../cms/sfe-adapter/sfe-metadata-wrapper';

@Component({
  selector: 'ish-cms-text',
  templateUrl: './cms-text.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class CMSTextComponent extends SfeMetadataWrapper {
  @Input() pagelet: ContentPageletView;
}
