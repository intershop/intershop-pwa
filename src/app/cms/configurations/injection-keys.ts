import { InjectionToken, SimpleChange, Type } from '@angular/core';

import { ContentPageletView } from 'ish-core/models/content-view/content-views';
import { SfeMetadata } from '../sfe-adapter/sfe.types';

export interface CMSComponentInterface {
  pagelet: ContentPageletView;
  setSfeMetadata(sfeInfo: SfeMetadata): void;
  ngOnChanges?(changes?: { pagelet: SimpleChange }): void;
}

export interface CMSComponentProvider {
  definitionQualifiedName: string;
  class: Type<CMSComponentInterface>;
}

export const CMS_COMPONENT = new InjectionToken<CMSComponentProvider>('InjectionToken');
