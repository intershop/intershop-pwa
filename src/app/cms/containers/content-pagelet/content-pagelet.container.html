<ng-container *ngIf="noMappingFound">
  <!-- default rendering for unknown pagelet definitions -->
  <div class="content-pagelet" style="border: 1px solid rgb(255, 255, 0); padding: 10px;">
    <div class="content-pagelet-info" [title]="pagelet.id">
      {{ pagelet.id }} ({{ pagelet.definitionQualifiedName }})
    </div>
  </div>
</ng-container>

<ng-container #cmsOutlet></ng-container>
