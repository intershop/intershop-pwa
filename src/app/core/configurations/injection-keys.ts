import { InjectionToken } from '@angular/core';

import { CookieConsentOptions } from 'ish-core/models/cookies/cookies.model';
import { ViewType } from 'ish-core/models/viewtype/viewtype.types';

import { environment } from '../../../environments/environment';

/**
 * Array of paths that always use mocked data
 */
export const API_MOCK_PATHS = new InjectionToken<string[]>('apiMockPaths', { factory: () => environment.apiMockPaths });

/**
 * If 'username' login name is used for registration, if 'email' the email is used as login name (default: 'email')
 */
export const USER_REGISTRATION_LOGIN_TYPE = new InjectionToken<string>('userRegistrationLoginType', {
  factory: () => 'email',
});

/**
 * The maximum subcategories level depth rendered in the main navigation
 */
export const MAIN_NAVIGATION_MAX_SUB_CATEGORIES_DEPTH = new InjectionToken<number>(
  'mainNavigationMaxSubCategoriesDepth',
  { factory: () => environment.mainNavigationMaxSubCategoriesDepth }
);

/**
 * global definition of the product listing page size
 */
export const PRODUCT_LISTING_ITEMS_PER_PAGE = new InjectionToken<number>('productListingItemsPerPage', {
  factory: () => environment.productListingItemsPerPage,
});

/**
 * default definition of the product listing view type
 */
export const DEFAULT_PRODUCT_LISTING_VIEW_TYPE = new InjectionToken<ViewType>('defaultProductListingViewType', {
  factory: () => environment.defaultProductListingViewType,
});

/**
 * the configured cookie consent options for the application
 */
export const COOKIE_CONSENT_OPTIONS = new InjectionToken<CookieConsentOptions>('cookieConsentOptions', {
  factory: () => environment.cookieConsentOptions,
});

/*
 * global definition of the Bootstrap grid system breakpoint widths
 */

export const MEDIUM_BREAKPOINT_WIDTH = new InjectionToken<number>('mediumBreakpointWidth', {
  factory: () => environment.mediumBreakpointWidth,
});

export const LARGE_BREAKPOINT_WIDTH = new InjectionToken<number>('largeBreakpointWidth', {
  factory: () => environment.largeBreakpointWidth,
});
