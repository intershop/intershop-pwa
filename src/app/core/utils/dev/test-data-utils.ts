import { CategoryTree, CategoryTreeHelper } from 'ish-core/models/category-tree/category-tree.model';
import { Category } from 'ish-core/models/category/category.model';
import { ContentPageTreeHelper } from 'ish-core/models/content-page-tree/content-page-tree.helper';
import { ContentPageTree, ContentPageTreeElement } from 'ish-core/models/content-page-tree/content-page-tree.model';

export function categoryTree(categories?: Category[]): CategoryTree {
  const tree = CategoryTreeHelper.empty();
  if (categories) {
    return categories.reduce((acc, cat) => CategoryTreeHelper.add(acc, cat), tree);
  }
  return tree;
}

export function pageTree(elements?: ContentPageTreeElement[]): ContentPageTree {
  const tree = ContentPageTreeHelper.empty();
  if (elements) {
    return elements.reduce((acc, cat) => ContentPageTreeHelper.add(acc, cat), tree);
  }
  return tree;
}
