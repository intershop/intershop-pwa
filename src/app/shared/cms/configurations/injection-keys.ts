import { InjectionToken, Type } from '@angular/core';

import { CMSComponent } from 'ish-shared/cms/models/cms-component/cms-component.model';

export interface CMSComponentProvider {
  definitionQualifiedName: string;
  class: Type<CMSComponent>;
}

export const CMS_COMPONENT = new InjectionToken<CMSComponentProvider>('InjectionToken');
