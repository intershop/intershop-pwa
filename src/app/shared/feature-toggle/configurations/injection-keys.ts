import { InjectionToken } from '@angular/core';

export const FEATURE_TOGGLES = new InjectionToken<{ [feature: string]: boolean }>('featureToggles');
