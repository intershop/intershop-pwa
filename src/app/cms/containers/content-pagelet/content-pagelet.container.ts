// tslint:disable:ccp-no-markup-in-containers
import {
  ChangeDetectionStrategy,
  Component,
  ComponentFactoryResolver,
  Injector,
  Input,
  OnChanges,
  SimpleChange,
  ViewChild,
  ViewContainerRef,
} from '@angular/core';

import { ContentPageletView } from 'ish-core/models/content-view/content-views';
import { SfeMapper } from '../../../cms/sfe-adapter/sfe.mapper';
import { CMSComponentInterface, CMSComponentProvider, CMS_COMPONENT } from '../../configurations/injection-keys';

@Component({
  selector: 'ish-content-pagelet',
  templateUrl: './content-pagelet.container.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ContentPageletContainerComponent implements OnChanges {
  @Input() pagelet: ContentPageletView;

  noMappingFound: boolean;

  private components: CMSComponentProvider[] = [];

  @ViewChild('cmsOutlet', { read: ViewContainerRef })
  cmsOutlet: ViewContainerRef;

  constructor(injector: Injector, private componentFactoryResolver: ComponentFactoryResolver) {
    this.components = injector.get<CMSComponentProvider[]>(CMS_COMPONENT, []);
  }

  ngOnChanges() {
    const mappedComponent = this.components.find(
      c => c.definitionQualifiedName === this.pagelet.definitionQualifiedName
    );

    if (mappedComponent) {
      const componentRef = this.createComponent(mappedComponent);
      this.initializeComponent(componentRef.instance);
    }

    this.noMappingFound = !mappedComponent;
  }

  private createComponent(mappedComponent: CMSComponentProvider) {
    const factory = this.componentFactoryResolver.resolveComponentFactory(mappedComponent.class);
    this.cmsOutlet.clear();
    return this.cmsOutlet.createComponent(factory);
  }

  private initializeComponent(instance: CMSComponentInterface) {
    instance.pagelet = this.pagelet;
    instance.setSfeMetadata(SfeMapper.mapPageletViewToSfeMetadata(this.pagelet));

    // OnChanges has to be manually invoked on dynamically created components
    if (instance.ngOnChanges) {
      instance.ngOnChanges({ pagelet: new SimpleChange(undefined, this.pagelet, true) });
    }
  }
}
