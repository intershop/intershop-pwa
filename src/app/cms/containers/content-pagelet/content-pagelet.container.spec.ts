import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed, async } from '@angular/core/testing';

import { ContentPagelet } from 'ish-core/models/content-pagelet/content-pagelet.model';
import { SafeHtmlPipe } from 'ish-core/pipes/safe-html.pipe';
import { createSimplePageletView } from 'ish-core/utils/dev/test-data-utils';
import { CMSTextComponent } from '../../components/cms-text/cms-text.component';
import { CMS_COMPONENT } from '../../configurations/injection-keys';

import { ContentPageletContainerComponent } from './content-pagelet.container';

describe('Content Pagelet Container', () => {
  let component: ContentPageletContainerComponent;
  let fixture: ComponentFixture<ContentPageletContainerComponent>;
  let element: HTMLElement;
  let pagelet: ContentPagelet;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [CMSTextComponent, ContentPageletContainerComponent, SafeHtmlPipe],
      providers: [
        {
          provide: CMS_COMPONENT,
          useValue: { definitionQualifiedName: 'fq-defined', class: CMSTextComponent },
          multi: true,
        },
      ],
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
    })
      .overrideComponent(ContentPageletContainerComponent, { set: { entryComponents: [CMSTextComponent] } })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ContentPageletContainerComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;
    pagelet = {
      definitionQualifiedName: 'fq',
      id: 'id',
      configurationParameters: {
        HTMLText: 'foo',
      },
    };
  });

  it('should be created', () => {
    expect(component).toBeTruthy();
    expect(element).toBeTruthy();
    component.pagelet = createSimplePageletView(pagelet);
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchSnapshot();
  });

  it('should render assigned template if name matches', () => {
    pagelet.definitionQualifiedName = 'fq-defined';
    component.pagelet = createSimplePageletView(pagelet);
    expect(() => fixture.detectChanges()).not.toThrow();
    expect(element).toMatchSnapshot();
  });
});
