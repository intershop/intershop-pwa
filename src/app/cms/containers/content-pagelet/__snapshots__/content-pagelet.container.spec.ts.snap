// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Pagelet Container should be created 1`] = `
<div
  id="root0"
  ng-version="7.1.4"
>

  <div />
</div>
`;

exports[`Content Pagelet Container should render assigned template if name matches 1`] = `
<div
  id="root1"
  ng-version="7.1.4"
>

  <div />
  <ish-cms-text
    data-sfe="{\\"id\\":\\"pagelet:domain:id\\",\\"displayName\\":\\"pagelet\\",\\"displayType\\":\\"defined\\",\\"renderObject\\":{\\"id\\":\\"id\\",\\"domainId\\":\\"domain\\",\\"type\\":\\"Pagelet\\"}}"
  >

    <span>
      foo
    </span>
  </ish-cms-text>
</div>
`;
